src/backend/opencl/program.cpp

```diff
@@ -67,4 +69,32 @@ namespace opencl
             throw;
         }
     }
+
+    NDRange getKernelLaunchCfg1D(const Kernel& kernel)
+    {
+        int device = getActiveDeviceId();
+
+        size_t dim0ThreadCount = getDeviceWorkGroupSize(kernel, device);
+
+        // make sure you are not launching what the device can't handle
+        dim0ThreadCount = std::min(getDevice(device).getInfo<CL_DEVICE_MAX_WORK_GROUP_SIZE>(), dim0ThreadCount);
+
+        return cl::NDRange(dim0ThreadCount, 1, 1);
+    }
+
+    NDRange getKernelLaunchCfg2D(const Kernel& kernel)
+    {
+        int device = getActiveDeviceId();
+
+        size_t dim0ThreadCount = kernel.getWorkGroupInfo<CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE>(getDevice(device));
+
+        size_t recomWorkGroupSize = getDeviceWorkGroupSize(kernel, device);
+
+        size_t dim1ThreadCount = divup(recomWorkGroupSize, dim0ThreadCount);
+
+        // make sure you are not launching what the device can't handle
+        dim1ThreadCount = std::min(getDevice(device).getInfo<CL_DEVICE_MAX_WORK_GROUP_SIZE>(), dim1ThreadCount);
+
+        return cl::NDRange(dim0ThreadCount, dim1ThreadCount, 1);
+    }
```

Comment:
I agree, this is not a single bullet for all cases. It is just like many other convenience functions we have internally. We can have a better name and cautionary comments before that function explaining where it should be used and what it does.

Don't you think the number returned by `clGetKernelWorkGroupInfo` for the param name `CL_KERNEL_WORK_GROUP_SIZE` takes into account the static `local` allocations declared in the kernel ?